Implemented Query Screen

AbhiKul10 opened the pull request:
**Describe the Feature you Added**
Designed and Implemented Query Screen

**How will your Feature Help?**
- Users can ask Queries
- Users can answer the Queries
- Image can be uploaded in a Query
- Like option for the answers (only available for the Author of Query)
- Changing of the Tag (Answer, Answered) based on-if the Question has been Answered
- Author of Query is able to delete the Query

Fixes #18  

<p align="center"  float="left">
<img src="https://user-images.githubusercontent.com/40908684/122526194-ac741700-d037-11eb-9ed3-fa1d3eac1ab1.gif" width="500" >

<img src="https://user-images.githubusercontent.com/40908684/122526268-bc8bf680-d037-11eb-9a3d-cb2ff4b708cc.gif" width="500" >
</p>

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,225 @@
+        })
+    }
+
+    iflikedButton(ansuid, lb) {
```
let's remove 'if' from these functions.

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,225 @@
+        })
+    }
+
+    UNSAFE_componentWillMount(){
```
This is deprecated as I remember. Let's use constructor or componentDidMount.

shehand commented on src/screens/AnswerScreen/styles.js:
```diff
@@ -0,0 +1,211 @@
+        borderWidth: 1,
+        backgroundColor: Platform.OS === 'ios' ? '#fff' : null,
+        borderColor: '#b8b8b8',
+        shadowColor: Platform.OS === 'ios' ? '#000000' : '#000000',
+        shadowOffset: {
+        width: Platform.OS === 'ios' ? 0 : 0,
+        height: Platform.OS === 'ios' ? 2 : 2,
+        },
```
Here I think we do not need the condition. Values are same.

shehand commented on src/screens/HomeScreen/HomeScreen.js:
```diff
@@ -1,19 +1,101 @@
+            var queryArray = []
+            querySnapshot.forEach((doc) => {
+                q_data = {
+                    "id": doc.data().id,
+                    "QueryInput": doc.data().QueryInput,
+                    "postTime": doc.data().postTime,
+                    "q_id": doc.id,
+                    "postdatetime": doc.data().postdatetime,
+                    "answer_present": doc.data().answer_present,
+                    "q_image" : doc.data().query_image
```
Can you see the issue here? Use one single way when declaring variables. Do not mix different methods as this will make other devs hard to understand the code .

AbhiKul10 replied:
Let me know if I changed the names correctly?

shehand commented on src/screens/HomeScreen/styles.js:
```diff
@@ -1,12 +1,126 @@
+        borderWidth: 1,
+        backgroundColor: Platform.OS === 'ios' ? '#fff' : null,
+        borderColor: '#b8b8b8',
+        shadowColor: Platform.OS === 'ios' ? '#000000' : '#000000',
+        shadowOffset: {
+        width: Platform.OS === 'ios' ? 0 : 0,
+        height: Platform.OS === 'ios' ? 2 : 2,
```
Same comment as previous.

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,168 @@
+          height: 300,
+          cropping: true,
+        }).then((image) => {
+          const imageUri = Platform.OS === 'ios' ? image.path : image.path;
```
Why condition?

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,168 @@
+          return
+        }
+          const uri = setimage;
+          const childPath = `post/${firebase.auth().currentUser.uid}/${Math.random().toString(36)}`;
```
Let's change this URI to post/uid/<unique id of the post>/random_string. This way I think the files be more organized. As a suggestion, let's have a variable listed in firebase that hold a integer value which will be the next post id. I think this can be declared in user profile data section. Value should ne updated once new post is added.

AbhiKul10 replied:
Okay! I'll change it, but I have no idea about the use of '//'. What is the use of it?

AbhiKul10 replied:
I didn't understand this approach.  Do you want me to add a new Integer field in User profile data (with Full name, email, interest, country) which will contain the post id (next?)? What it will be used for?

shehand replied:
Sorry, I missed what I meant here. I have a question, when we give answerers, are they also posted into the same object that is created in the post/uid URI?

shehand commented on src/screens/QuestionScreen/styles.js:
```diff
@@ -0,0 +1,130 @@
+        borderWidth: 1,
+        backgroundColor: Platform.OS === 'ios' ? '#fff' : null,
+        borderColor: '#b8b8b8',
+        shadowColor: Platform.OS === 'ios' ? '#000000' : '#000000',
+        shadowOffset: {
+        width: Platform.OS === 'ios' ? 0 : 0,
+        height: Platform.OS === 'ios' ? 2 : 2,
```
Same comment

shehand reviewed (changes requested):
Overall, the feature is fine. I think when uploading the images and adding the answers, the alert message is not necessary. The reason is there is a visual output that user can see and understand that of what he/she did. But there should be alert message for errors.

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,224 @@
+                firebase.firestore().collection("users").doc(query.id).onSnapshot((userObject) => {
+                    a_data = {
+                        "answer_uid" : query.Answer_id,
+                        "and_time" : query.AnsTime,
```
and or ans?

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,224 @@
+                    a_data = {
+                        "answer_uid" : query.Answer_id,
+                        "and_time" : query.AnsTime,
+                        "q_ans" : query.AnswerQuery,
```
make q as query, q makes no sense here

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,224 @@
+                        <TouchableOpacity 
+                        style={styles.AnswerButtonStyle}
+                        onPress={() => {
+                            var timeDate = moment()
+                            const {AnswerQuery} = this.state
+                            if(AnswerQuery && AnswerQuery.length > 0){
+                            firebase.firestore().collection("queries").doc(this.state.data).collection("answers").add({
+                                id: firebase.auth().currentUser.uid,
+                                AnswerQuery,
+                                AnsTime: timeDate.format('lll'),
+                                LikeButton,
+                                ans_data_time: firebase.firestore.FieldValue.serverTimestamp()
+                            }).then((docRef) => {
+                                this.setState({
+                                    AnswerQuery: ''
+                            })
+                                firebase.firestore().collection("queries").doc(this.state.data).collection("answers").doc(docRef.id).update({
+                                    Answer_id: docRef.id
+                                });
+                            })
+                            .catch((error) => {
+                                alert("Error occured while adding your Answer")
+                            });
+                        } else {
+                            alert("Please Enter Valid Answer")
+                            }
+                        }}
```
move out from the render method and create a separate function

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,224 @@
+        this.getData()
+    }
+
+    render() {
+        const { answerdetail, Questionname, QuestionAskid, LikeButton, q_image, Questionid } = this.state
+        user_uid = firebase.auth().currentUser.uid
+    return(
+        <SafeAreaView>
```
alignment issue

shehand commented on src/screens/HomeScreen/HomeScreen.js:
```diff
@@ -1,19 +1,101 @@
+            var queryArray = []
+            querySnapshot.forEach((doc) => {
+                q_data = {
+                    "uid": doc.data().id,
+                    "QueryInput": doc.data().QueryInput,
+                    "postTime": doc.data().postTime,
+                    "id": doc.id,
+                    "postdatetime": doc.data().postdatetime,
+                    "answerpresent": doc.data().answer_present,
+                    "queryimage" : doc.data().query_image
```
As I said, if you are using lower case _ method, use it everywhere. Make that as a standard for this project. Code quality is a must when it comes to OSS.

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+class QuestionScreen extends  React.Component {
+    state = {QueryInput : '', setimage : null, setDisplayImage: null, transferred: 0};
+
+    takephotofrommlib = () => {
```
See the method names here, three methods are having three different naming convention methods. I did not go this deeper in other PR review, thought you might have familiar about these.

AbhiKul10 replied:
I'll keep that in mind! Changed variable and method naming convention to camelCase type.

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+                
+            <TouchableOpacity 
+            onPress={() => {
+                var timeDate = moment()
+                const Useruid = firebase.auth().currentUser.uid
+                const {QueryInput} = this.state
+                if(QueryInput && QueryInput.length > 0){
+                firebase.firestore().collection("queries").add({
+                    id: Useruid,
+                    QueryInput,
+                    postdatetime: firebase.firestore.FieldValue.serverTimestamp(),
+                    postTime: timeDate.format('lll'),
+                    answer_present: false,
+                    query_image: setDisplayImage
+                })
+                .then((docRef) => {
+                    this.setState({
+                        QueryInput: '',
+                        setDisplayImage: null,
+                        setimage: null
+                    })
+                })
+                .catch((error) => {
+                    alert("Error occured while adding your Query")
+                });
+            } else {
+                alert("Please Enter Valid Query")
+            }   
```
Move outside the render method

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+            ]  
+        );  
+      }
+    render() {
+        const {setDisplayImage, setimage, transferred} = this.state
+    return(
+    <SafeAreaView>
+        <View>
```
Alignment issues

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+          return
+        }
+          const uri = setimage;
+          const childPath = `post/${firebase.auth().currentUser.uid}//${Math.random().toString(36)}`;
```
Sorry, I missed what I meant here. I have a question, when we give answerers, are they also posted into the same object that is created in the post/uid URI? Change this back to original what you had earlier.

AbhiKul10 replied:
No! When a user gives an answer, A new collection is created in that particular Query Collection. But the image is saved in firebase storage, whose URL is saved in this answer collection. `childPath` is the URI of that image in firebase storage and `uri` is the image location in local storage

<img width="1260" alt="Screenshot 2021-06-19 at 3 06 47 PM" src="https://user-images.githubusercontent.com/40908684/122638143-211a8480-d110-11eb-97b2-c56fb79b7401.png">

AbhiKul10 replied:
Every image when posted will be saved in a location with the format of childPath. So every image posted by a particular user in this application will be saved in a folder with his uid. Here is the path of images in firebase storage in below screenshot

<img width="1135" alt="Screenshot 2021-06-19 at 3 17 11 PM" src="https://user-images.githubusercontent.com/40908684/122638437-b23e2b00-d111-11eb-8adc-0921e38e19d1.png">

shehand replied:
Oh OK, in that case shall we change the URL to post/uid/queries/random_string. That would organize the files more convenient way I think. Better if you can organize the images/file based by queries. Let me know what you are going to do here then.

Wathsara replied:
+1 @shehand. Better to organize the files in a better way.

AbhiKul10 replied:
Yes! I can change the URL to `post/uid/queries/random_string`, but changing it to `post/uid/queries/post_id/random_string` is not possible, because at first, I upload the image in firebase storage, and then we post the query. At the time of uploading of the image, we don't have post_id. It creates when we post the query. Let me know if changing it to `post/uid/queries/random_string` is fine.

shehand replied:
That's where you come up with a creative approach. You can do it when a variable is declared for each user which hold the next valid post id. This value is incremented by one each time new post is posted by the user. There can be situation when providing answers, need to add images. In that case we can add those kind of images to the same folder and this will make the storage kind of organized(Note: nothing is impossible, you just have to think out of the box). Let us know you got the idea what we are referring here.

AbhiKul10 replied:
Okay! I Got it. Let me Brainstorm the idea and come up with a good solution

AbhiKul10 replied:
Here is the Screenshot of the path of the image.

<img width="1143" alt="Screenshot 2021-06-20 at 7 43 52 PM" src="https://user-images.githubusercontent.com/40908684/122677506-31119180-d200-11eb-8380-2fb8c3057e82.png">


What I did is- I created a new field in the user section `postNumber` which will be incremented by one every time a user post a query. So the path is `post/uid/queries/postNumber/randomString`. `postNumber` field can also be used as a post count by an individual user.

@shehand I guess this is what you wanted me to implement!? Let me know if this is fine.

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+        const {setImage} = this.state
+        this.setState({transferred: 0})
+        ImagePicker.openPicker({
+          width: 300,
+          height: 300,
```
As I said before, hope these will not affect to different devices(better to check with some tablets also since they are also listed in mobile devices category)

AbhiKul10 replied:
I'll remove the cropping option from this picker. This height and width are only valid when cropping of an image is enabled. Removing cropping makes height and width variable according to the device.

shehand replied:
I'm not referring to this specific scenario. I'm talking about the whole styles that are defined without device height/width variables.

AbhiKul10 replied:
Updated Dimensions

shehand commented on src/screens/QuestionScreen/QuestionScreen.js:
```diff
@@ -0,0 +1,167 @@
+        if(queryInput && queryInput.length > 0){
+            firebase.firestore().collection("queries").add({
+                id: Useruid,
+                queryInput,
```
Isn't there should be a key here?

AbhiKul10 replied:
Do you mean key-value pair? Using this like `queryInput: queryInput` is same as `queryInput` I guess. The data added through this line gives me the same output as with key: value pair. Let me know if this approach is fine or should I change it?

shehand replied:
Can you debug and see what is the value of the passing queryInput? I'm not confident on this approach however. If you say this is working, that's fine. But better to verify with debugging the value.

AbhiKul10 replied:
I double-checked it using the remote debugger, It has the same value that I passed in using textinput. I guess this may lead to confusion in the future so I should add a key to it!

shehand reviewed (approved):
Will merge after @AbhiKul10  addresses the comments.

AbhiKul10 commented:
I will be implementing Image upload in the answer too! Please merge after that commit

AbhiKul10 commented:
I have Added Image upload feature in the Answer Screen too. User can upload image when answering a query. Updated path for the answer images in firebase storage as `post/authoruid/queries/postNumber/answers/${Math.random().toString(36)}`

Please review @shehand @Wathsara 

**Answer Image path**
<img width="1142" alt="Screenshot 2021-06-21 at 7 33 47 PM" src="https://user-images.githubusercontent.com/40908684/122775512-d5aad680-d2c7-11eb-8d0e-fabeeb8e668e.png">

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,293 @@
+
+class AnswerScreen extends React.Component {
+    userUid = firebase.auth().currentUser.uid
+    state = {answerQuery : '', answerDetail: [], questionName: '', questionAskId : '', likeButton: false, queryImage: null, questionId: '', transferred: 0, setImage: null, setDisplayImage: null, answerImage: null, answerPostNumber: null};
```
Normally the maximum characters that would be in a sing line of code is 120 so that is easy to review and no need to scroll horizontally.

shehand commented on src/screens/AnswerScreen/AnswerScreen.js:
```diff
@@ -0,0 +1,293 @@
+                                            <View>
+                                                <Text style={styles.Q_ansStyle}>{item.answerQuery}</Text>
+                                                { item.answerImage ? 
+                                                <TouchableOpacity onPress={() => Linking.openURL(item.answerImage)}><Image source={{uri: item.answerImage}} style={styles.ImageStyle1} /></TouchableOpacity> : null}
```
set character limit for a single line(including spaces/tabs also)

AbhiKul10 replied:
Updated👍🏻

shehand reviewed (approved):
Quick fix this character length problem so that we can close this. Nice work by the way @AbhiKul10!
